types/react-query/index.d.ts

```diff
@@ -36,107 +53,204 @@ export interface QueryOptions<TResult> {
+    onSettled?: (data: TResult, error: any) => Promise<any> | undefined;
+    onError?: (error: any) => Promise<any> | undefined;
+    throwOnError?: boolean;
+    useErrorBoundary?: boolean;
 }
 
 export type MutateFunction<TResults, TVariables extends object> = (
     variables?: TVariables,
     options?: {
-        updateQuery?: string | [string, object],
+        updateQuery?: string | [string, object];
         waitForRefetchQueries?: boolean;
-    }
+    },
 ) => Promise<TResults>;
 
-export interface MutationResult<TResults> {
-    data: TResults | null;
-    isLoading: boolean;
+export interface MutationState<TResults> {
+    status: 'idle' | 'loading' | 'error' | 'success';
+    data: TResults | undefined;
     error: null | Error;
     promise: Promise<TResults>;
-    reset: () => void;
 }
 
-export function setQueryData(
-    queryKey: string | [string, object],
-    data: any,
-    options?: {
-        shouldRefetch?: boolean
-    }
-): void | Promise<void>;
-
-export function refetchQuery(
-    queryKey: string | [string, object] | [string, false],
-    force?: {
-        force?: boolean
-    }
-): Promise<void>;
-
-export function refetchAllQueries(
-    options?: {
-        force?: boolean,
-        includeInactive: boolean
-    }
-): Promise<void>;
+export namespace queryCache {
+    function prefetchQuery<TResult, TQueryKey extends QueryKey>(
+        queryKey: TQueryKey,
+        queryFn: (key: string, ...args: QueryKeyVariablesArgs<TQueryKey>) => Promise<TResult>,
+        options?: QueryOptions<TResult>,
+    ): Promise<TResult>;
+
+    // tslint:disable:no-unnecessary-generics
+    function getQueryData<TResult>(queryKey: QueryKey): TResult | undefined;
+
+    function setQueryData<TResult>(queryKey: QueryKey, updater: TResult | ((oldData: TResult) => TResult)): void;
+
+    function refetchQueries(
+        queryKeyOrPredicateFn: QueryKey | ((query: QueryKey) => boolean),
+        options?: {
+            exact?: boolean;
+            throwOnError?: boolean;
+        },
+    ): Promise<void>;
+
+    function removeQueries(
+        queryKeyOrPredicateFn: QueryKey | ((query: QueryKey) => boolean),
+        options?: {
+            exact?: boolean;
+        },
+    ): void;
+
+    // TODO: type returned QueryObject
+    function getQuery(queryKey: QueryKey): any;
+
+    const isFetching: boolean;
+
+    function subscribe(callback: (queryCache: any) => void): (unsubscribe: (...args: any[]) => void) => void;
+
+    function clear(): void;
+}
 
 export function useIsFetching(): boolean;
 
 export const ReactQueryConfigProvider: React.ComponentType<{
-    config?: ReactQueryProviderConfig
+    config?: ReactQueryProviderConfig;
 }>;
 
 export interface ReactQueryProviderConfig {
+    suspense?: boolean;
+    useErrorBoundary?: undefined;
+    throwOnError?: boolean;
+    refetchAllOnWindowFocus?: boolean;
+    queryKeySerializerFn?: (queryKey: QueryKey) => [string, any[]];
+    onSuccess?: () => {};
```

Comment:
```suggestion
    onSuccess?: (data: object) => data;
```
This should have data as the parameter and return the same data. I'm not sure how we can type the data.

Reply:
`onSuccess` callback must not return the same data - even the default implementation is a `noop` fn (https://github.com/tannerlinsley/react-query/blob/master/src/config.js#L19).  

```
onSuccess?: (data: unknown) => Promise<any> | void
```
should be the correct signature